modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExecutionServiceImpl.java

```diff
@@ -787,6 +790,13 @@ public CompletableFuture<Void> closeAsync() {
         }
 
         private void enlistPartitions(MappedFragment mappedFragment, InternalTransaction tx) {
+            // It's possible to have null txState if transaction isn't started yet.
+            if (tx != null && !(tx.state() == TxState.PENDING || tx.state() == null)) {
```

Comment:
what if transaction will be finished right after we pass this check?

Reply:
It will propagate deeper and will be considered a new one (different from this transaction). But I know what you mean: "the new transaction will have the same ID".
Unfortunately, we did not resolve the issue before because we thought it had been resolved on the server side. But it does not work as expected.
Maybe we should review our decision....

Reply:
The code was deleted. The new logic is based on rw-lock in the transaction enlist method.